### PR TITLE
NOISSUE - Twins page

### PR DESCRIPTION
**Pull request by darkodraskovic**
Add twins management page to admin section.

### CODE REVIEW

File: src/app/common/interfaces/dbreader.interface.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+export interface DbReaderMetadata {
+  type?: string;
+  db_reader_data: any;
```
Comment (manuio) [inline src/app/common/interfaces/dbreader.interface.ts]: Use camel case

File: src/app/common/services/dbreader/dbreader.service.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+import { ThingsService } from 'app/common/services/things/things.service';
+import { ChannelsService } from 'app/common/services/channels/channels.service';
+import { NotificationsService } from 'app/common/services/notifications/notifications.service';
+import { DbReaderNode } from 'app/common/interfaces/dbreader.interface';
+
+@Injectable()
+export class DbReaderService {
+  typeDbReader = 'dbReader';
+  typeDbReaderServer = 'dbReader';
```
Comment (manuio) [inline src/app/common/services/dbreader/dbreader.service.ts]: Why this repetition?

File: src/app/common/services/dbreader/dbreader.service.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      name: node.name,
+      metadata: {
+        type: this.typeDbReader,
+        channelID: '',
+        db_reader_data: node.db_reader_data,
+      },
+    };
+
+    // Check if a channel exist for serverURI
+    return this.channelsService.getChannels(0, 1, 'dbReader', `{"serverURI": "${node.serverURI}"}`).map(
```
Comment (manuio) [inline src/app/common/services/dbreader/dbreader.service.ts]: Use `typeDbReader` here and fix the metadata query

File: src/app/common/services/twins/twins.service.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+          return Observable.throw(err);
+        },
+      );
+  }
+
+  deleteTwin(twinID: string) {
+    return this.http.delete(environment.twinsUrl + '/' + twinID)
+      .map(
+        resp => {
+          this.notificationsService.success('Device successfully deleted', '');
```
Comment (manuio) [inline src/app/common/services/twins/twins.service.ts]: `Twin successfully deleted`

File: src/app/pages/admin/twins/details/twins.details.component.html [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+          <div class="col-2"> Thing ID: </div><div class="col-5">{{twin.thing_id}}</div>
+          <div class="col-5">
+            <nb-select placeholder="Select thing" [(selected)]="selectedThing" size="tiny">
+              <nb-option *ngFor="let thing of this.things" [value]="thing">{{thing.name}}</nb-option>
+            </nb-select>
+          </div>
+          
+          <div class="col-2"> Created: </div><div class="col-10"> {{ twin.created }} </div>
+          
+          <div class="col-2"> Revision: </div><div class="col-10"> {{ twin.revision }} </div>
```
Comment (manuio) [inline src/app/pages/admin/twins/details/twins.details.component.html]: Return a new line between columns. Make it clearer please

File: src/app/pages/admin/twins/details/twins.details.component.scss [hunk 6]
```diff
@@ -0,0 +1,27 @@
+nb-icon {
+  vertical-align: middle;
+}
+
+.col-2 {
+  /* text-align: right; */
+  /* padding: 0; */
```
Comment (manuio) [inline src/app/pages/admin/twins/details/twins.details.component.scss]: Dead code

File: src/app/pages/things/dbreader/templates/mssql/mssql.component.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  ngOnInit() {
+    this.eventsSubscription = this.events.subscribe(() => this.onEdit());
+  }
+  onEdit() {
+    if (this.metadataForm.valid) {
+
+      if (this.thing.metadata !== '') {
+        try {
+          // this.thing.metadata = JSON.parse(this.thing.metadata);
```
Comment (manuio) [inline src/app/pages/things/dbreader/templates/mssql/mssql.component.ts]: dead code

File: src/app/pages/things/dbreader/templates/mssql/mssql.component.scss [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  .error {
+      color: red;
+      font-size: smaller;
+  }
+  
+  label {
+    font-size: small;
+  }
```
Comment (manuio) [inline src/app/pages/things/dbreader/templates/mssql/mssql.component.scss]: Return new line here

File: src/app/pages/things/dbreader/templates/mssql/mssql.component.html [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            <div class="col-md-12 decorate">
+                <label for="interval">Interval(minutes):</label>
+            </div>
+            <div class="col-md-3 decorate">              
+                <input type="number" nbInput id="interval" fullWidth status="primary" fieldSize="small" [placeholder]="thing.metadata.db_reader_data.interval" [(ngModel)]="thing.metadata.db_reader_data.interval" name="interval"
+                #interval="ngModel">     
+            </div>
+        </div>
+    </div>
+    <!-- <input type="text" hidden nbInput status="primary" fullWidth fieldSize="small" [placeholder]="thing.metadata.db_reader_data.dbtype" [(ngModel)]="thing.metadata.db_reader_data.dbtype" required>   -->
```
Comment (manuio) [inline src/app/pages/things/dbreader/templates/mssql/mssql.component.html]: dead code

File: src/app/pages/things/dbreader/details/dbreader.details.component.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    this.thingsService.connectedChannels(this.thing.id).subscribe(
+      (respConns: any) => {
+        this.connections = respConns.channels;
+        this.channelsService.getChannels(this.offset, this.limit).subscribe(
+          (respChans: any) => {
+            respChans.channels.forEach(chan => {
+              if (!(this.connections.filter(c => c.id === chan.id).length > 0)) {
+                this.channels.push(chan);
+              } else {
```
Comment (manuio) [inline src/app/pages/things/dbreader/details/dbreader.details.component.ts]: What else?

File: src/app/pages/things/dbreader/details/dbreader.details.component.scss [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+textarea {
+  width: 100%;
+  height: 100%;
+}
+
+label {
+  font-size: small;
+}
```
Comment (manuio) [inline src/app/pages/things/dbreader/details/dbreader.details.component.scss]: Return a new line

File: src/app/pages/things/dbreader/details/dbreader.details.component.html [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        </div>
+        <div *ngIf="connections.length !== 0">
+          <div *ngFor="let con of connections;" class="row" style="padding: 2.6px;">
+            <div class="col-9">
+              <nb-icon status="info" icon="arrow-right-outline"></nb-icon>
+              {{con.name}}
+            </div>
+            <div class="col-3">
+            
+            </div>
```
Comment (manuio) [inline src/app/pages/things/dbreader/details/dbreader.details.component.html]: What is this?

File: src/app/pages/things/dbreader/dbreader.component.ts [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+  onEditConfirm(event): void {
+       // close edditable row
+       event.confirm.resolve();
+
+       this.thingsService.editThing(event.newData).subscribe(
+         resp => {
+           this.notificationsService.success('DB Reader successfully edited', '');
+         },
+       );
+
```
Comment (manuio) [inline src/app/pages/things/dbreader/dbreader.component.ts]: Remove this blank line

File: src/app/pages/things/dbreader/dbreader.component.scss [hunk 14]
```diff
@@ -0,0 +1,8 @@
+button {
+    margin: 2px;
+  }
+  
+  nb-card-header {
+    text-align: center;
+  }
```
Comment (manuio) [inline src/app/pages/things/dbreader/dbreader.component.scss]: Fix tabulation

File: src/app/pages/things/dbreader/dbreader.component.html [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  
+    <nb-card-body>
+      <ng2-smart-table [settings]="settings"
+                       [source]="source"
+                       (createConfirm)="onCreateConfirm($event)"
+                       (editConfirm)="onEditConfirm($event)"
+                       (deleteConfirm)="onDeleteConfirm($event)">
+      </ng2-smart-table>
+    </nb-card-body>
+  </nb-card>
```
Comment (manuio) [inline src/app/pages/things/dbreader/dbreader.component.html]: Return a new line

File: src/app/pages/admin/twins/twins.component.ts [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    },
+    columns: {
+      id: {
+        title: 'ID',
+        editable: false,
+        addable: false,
+        filter: false,
+      },
+      thing_id: {
+        title: 'Thing ID',
```
Comment (manuio) [inline src/app/pages/admin/twins/twins.component.ts]: Thing or Twin here?

File: src/app/pages/admin/twins/twins.component.ts [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        type: 'string',
+        editable: false,
+        placeholder: 'Search thing id',
+        filter: {
+          placeholder: 'Search thing id',
+        },
```
Comment (manuio) [inline src/app/pages/admin/twins/twins.component.ts]: This have probably no effect

File: src/app/pages/admin/twins/twins.component.ts (hunk 16 above)
Comment (darkodraskovic) [reply src/app/pages/admin/twins/twins.component.ts]: Thing ID. It's a thing connected to a twin.

**Pull request merged**